cpp/lib/src/ZoteroHarvesterConversion.cc

```diff
@@ -531,61 +533,90 @@ std::string TikaDetectLanguage(const std::string &record_text) {
 }
 
 
-void DetectLanguage(MetadataRecord * const metadata_record, const Config::JournalParams &journal_params) {
-    // Normalize given language
-    if (not Config::IsAllowedLanguage(metadata_record->language_)) {
-        LOG_WARNING("Removing invalid language: " + metadata_record->language_);
-        metadata_record->language_.clear();
-    } else if (not Config::IsNormalizedLanguage(metadata_record->language_)) {
-        const std::string normalized_language(Config::GetNormalizedLanguage(metadata_record->language_));
-        LOG_DEBUG("Normalized language: " + metadata_record->language_ + " => " + normalized_language);
-        metadata_record->language_ = normalized_language;
+void NormalizeGivenLanguages(MetadataRecord * const metadata_record) {
+    // Normalize given languages
+    // We cant remove during iteration, so we use a multipass approach
+    std::set<std::string> unallowed_languages;
+    std::set<std::string> normalized_languages;
+    for (const auto &language : metadata_record->languages_) {
+        if (not Config::IsAllowedLanguage(language)) {
+            LOG_WARNING("Removing invalid language: " + language);
+            unallowed_languages.emplace(language);
+        } else if (not Config::IsNormalizedLanguage(language)) {
+            const std::string normalized_language(Config::GetNormalizedLanguage(language));
+            LOG_INFO("Normalized language: " + language + " => " + normalized_language);
+            unallowed_languages.emplace(language);
+            normalized_languages.emplace(normalized_language);
+        }
+    }
+    for (const auto &language : unallowed_languages)
+        metadata_record->languages_.erase(language);
+    for (const auto &normalized_language : normalized_languages)
+        metadata_record->languages_.emplace(normalized_language);
+}
```

Comment:
Wouldn't it be more straighftorward to collect the results in a temporary object where the unallowed languages are just skipped and assign this object at the end?

Reply:
You're right, the code is really more complicated than it could be... i will simplify it. However, instead of copying the whole MetadataRecord, i'd rather prefer to copy only the languages_ property, so the amount of data touched is minimized to keep it efficient and as safe as possible (e.g. to avoid data loss if there is any bug in the copy constructor)